### PR TITLE
Move some Loader request options into Loader options

### LINKED ISSUE #4880: Improve Loader policy specification/move stateful Container options into Loader policy
Followup from issue #4489, pull #4862

There are some stateful options (`LoaderHeader.cache`, `LoaderHeader.pause`, `LoaderHeader.reconnect`, and forceReadonly in particular) that can be provided in a Loader request that get applied in some way when the Container is created.  These options should not vary per Container, and should be consistent across all Containers created by a Loader.  If different options are needed, a new Loader should be created first.  This is particularly relevant in the guest component case, where a container will load another container.  The loading container does not know about the host's policies, but they should be applied by default to the loaded container.  To that end, these options should be moved into the Loader policy from request headers, and consumers should be updated to create a new Loader where needed.

### PR DESCRIPTION
**Pull request by heliocliu**
Fixes #4880

Add some Loader option equivalents for some Loader request options that are static for each Loader instance. This better defines how a Loader handles requests and better supports the guest component scenario.  Callers who need to change those Loader options should create a new Loader.

Also add an event for loading a container on the loader that the host can use to apply any additional policy not captured by loader options.

Future: executionContext should probably move to the Loader options as well, but its usage is a bit murky at the moment so I'll defer that to the background summarization issue.

### CODE REVIEW

File: packages/loader/container-definitions/src/loader.ts [hunk 1]
```diff
@@ -172,6 +172,13 @@ export interface IContainer extends IEventProvider<IContainerEvents>, IFluidRout
     request(request: IRequest): Promise<IResponse>;
 }
 
+/**
+ * Events emitted by the Loader "upwards" to the Host
+ */
+export interface ILoaderEvents extends IEvent {
```
Comment (anthony-murphy) [inline packages/loader/container-definitions/src/loader.ts:178]: literally just filed an issue for this: #5114

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -194,19 +194,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
                 resolvedUrl,
-                canReconnect: !(request.headers?.[LoaderHeader.reconnect] === false),
+                canReconnect,
             });
 
         return PerformanceEvent.timedExecAsync(container.logger, { eventName: "Load" }, async (event) => {
             return new Promise<Container>((res, rej) => {
                 const version = request.headers?.[LoaderHeader.version];
-                const pause = request.headers?.[LoaderHeader.pause];
+                const pause = loader.services.options.pause === true || request.headers?.[LoaderHeader.pause];
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:211]: i like to log a warning when something deprecated is used, so clients know they need to change something

File: packages/loader/container-loader/src/loader.ts [hunk 3]
```diff
@@ -396,6 +398,8 @@ export class Loader extends EventEmitter implements ILoader {
                     resolvedAsFluid);
         }
 
+        this.emit("containerLoaded", container, newContainer);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/loader.ts:401]: >emit [](start = 13, length = 4)

i don't think we should emit when getting some cache, only when newly created

**Comment by anthony-murphy** [comment]
>     [LoaderHeader.cache]: boolean;

deprecate them here as well

---
Refers to: packages/loader/container-definitions/src/loader.ts:281 in 7e13f18. [](commit_id = 7e13f188e13a285eb92244cc635236e32c8ebdbe, deletion_comment = False)

File: packages/loader/container-definitions/src/loader.ts [hunk 4]
```diff
@@ -227,6 +246,7 @@ export type ILoaderOptions = {
 export enum LoaderHeader {
     /**
      * Override the Loader's default caching behavior for this container.
+     * @deprected 0.35 Use IloaderOptions.cache instead
```
Comment (anthony-murphy) [inline packages/loader/container-definitions/src/loader.ts:249]: how will the summarizer work?

**Comment by anthony-murphy** [comment]
> function canUseCache(request: IRequest): boolean {

seems like we are inconsistent here. probably want to remove this method, and use this.canUseCache

---
Refers to: packages/loader/container-loader/src/loader.ts:44 in 7e13f18. [](commit_id = 7e13f188e13a285eb92244cc635236e32c8ebdbe, deletion_comment = False)

File: packages/loader/container-definitions/src/loader.ts (hunk 1 above)
Comment (heliocliu) [reply packages/loader/container-definitions/src/loader.ts:178]: you commented the same in the issue i'm resolving against so i made sure to add it 😁

File: packages/loader/container-loader/src/loader.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/loader.ts:401]: i think we may just want to not do this change in this PR. there is some good discussion in #5114, and adding this the wrong way could make it harder to trim/remove the loader from the container runtime

**Comment by heliocliu** [comment]
Closing for the reasons in the attached issue.  The "stateful options" we would be moving in along with their new usage pattern don't fit well with how they're currently used, and the arbitrary policy callback isn't being added here either

**Pull request closed without merging**